platform/msm8916/msm8916-clock.c

```diff
@@ -684,14 +684,150 @@ static struct clk_lookup msm_clocks_8916[] =
+#define APCS_ALIAS0_CORE_CBCR_REG			0x8
+#define APCS_ALIAS0_CORE_CBCR_SW_CTL			BIT(0)
+
+static void msm8939_clock_c0_enable_sr2_pll(void)
+{
+	uint32_t base = APCS_C0_PLL_BASE;
+	int regval;
+
+	/* Disable Mode */
+	writel(0, base + C0_PLL_MODE_REG);
+
+	/* Configure L/M/N values - these are accumulator fields */
+	writel(52, base + C0_PLL_L_VAL_REG);
+	writel(0,  base + C0_PLL_M_VAL_REG);
+	writel(1,  base + C0_PLL_N_VAL_REG);
+
+	/* Configure USER_CTL and CONFIG_CTL value */
+	writel(C0_PLL_USER_CTL_MN_EN |
+	       C0_PLL_USER_CTL_PLLOUT_LV_EARLY |
+	       C0_PLL_USER_CTL_PLLOUT_LV_BIST |
+	       C0_PLL_USER_CTL_PLLOUT_LV_AUX |
+	       C0_PLL_USER_CTL_PLLOUT_LV_MAIN,
+	       base + C0_PLL_USER_CTL_REG);
+
+	/*
+	 * The downstream write value is 0x4c015765 which sets several reserved
+	 * bits to unknown values
+	 * 0x4c015765 => 1 0 0 11 00 0 0 00 00 01 0 1 01 0 11101100101
+	 * RESERVE_BIT_30 : 1
+	 * FORCE_PFD_UP |29| 0
+	 * FORCE_PFD_DOWN |28| 0
+	 * NMOSC_FREQ_CTRL |27:26| 11
+	 * PFD_DZSEL |25:24| 00
+	 * NMOSC_EN |23| 0
+	 * RESERVE_BIT22 |22| 0
+	 * ICP_DIV |21:20| 00
+	 * IREG_DIV |19:18| 00
+	 * CUSEL |17:16| 01
+	 * REF_MODE |15| 0
+	 * RESERVE_BIT14 |14| 1
+	 * CFG_LOCKDET |13:12| 01
+	 * FORCE_ISEED |11| 0
+	 * RESERVE_BITS10_0 |0:10| 11101100101
+	 */
+	writel(C0_PLL_CONFIG_CTL_RESERVE_BIT_30	|
+	       C0_PLL_CONFIG_CTL_NMOSC_FREQ_CTRL_PFD_DZSEL_4X |
+	       C0_PLL_CONFIG_CTL_CUSEL_250FF |
+	       C0_PLL_CONFIG_CTL_RESERVE_BIT14 |
+	       C0_PLL_CONFIG_CTL_CFG_LOCKDET_2PI15 |
+	       C0_PLL_CONFIG_CTL_RESERVE_BITS10_0_MAGIC, base + C0_PLL_CONFIG_CTL_REG);
+
+	/* Enable PLL now */
+	writel(C0_PLL_MODE_PLL_BYPASSNL, base + C0_PLL_MODE_REG);
+	udelay(2);
+	writel(C0_PLL_MODE_PLL_RESET_N |
+	       C0_PLL_MODE_PLL_BYPASSNL, base + C0_PLL_MODE_REG);
+	udelay(50);
+	writel(C0_PLL_MODE_PLL_RESET_N |
+	       C0_PLL_MODE_PLL_BYPASSNL |
+	       C0_PLL_MODE_PLL_OUTCTRL,
+	       base + C0_PLL_MODE_REG);
+}
+
+static void msm8939_clock_c0_init(void)
+{
+	uint32_t base;
+	int regval;
+	int count;
+
+	msm8939_clock_c0_enable_sr2_pll();
+
+	base = APCS_ALIAS0_CMD_RCGR_BASE;
+
+	/* Source GPLL0 and 1/2 the rate of GPLL0 0x0403 */
+	regval = APCS_ALIAS0_CFG_RCGR_SRC_SEL_CLK3 |
+		 APCS_ALIAS0_CFG_RCGR_SRC_DIV_2;
+	writel(regval, base + APCS_ALIAS0_CFG_REG);
+	regval = readl(base + APCS_ALIAS0_CFG_REG);
+
+	/* update bit */
+	regval = readl(base + APCS_ALIAS0_CMD_RCGR_REG);
+	regval |= APCS_ALIAS0_CMD_RCGR_UPDATE;
+	writel(regval, base + APCS_ALIAS0_CMD_RCGR_REG);
+
+	/* Wait for update to take effect */
+	for (count = 500; count > 0; count--) {
+		regval = readl(base + APCS_ALIAS0_CMD_RCGR_REG);
+		if (!regval & APCS_ALIAS0_CMD_RCGR_UPDATE)
+			break;
+		udelay(1);
+	}
+
+	/* Enable the branch */
+	regval =  readl(base + APCS_ALIAS0_CORE_CBCR_REG);
+	regval |= APCS_ALIAS0_CORE_CBCR_SW_CTL;
+	writel(regval, base + APCS_ALIAS0_CORE_CBCR_REG);
+	regval = readl(base + APCS_ALIAS0_CORE_CBCR_REG);
```

Comment:
Exactly the same register is already written during the L2 power up sequence:

```c
	/* Enable clocks via SW_CLK_EN */
	writel(APCS_L2_CORE_CBCR_CLK_ENABLE, base + APCS_L2_CORE_CBCR);
```

I don't think this should be here. :)

Reply:
I think it's fine to have the configuration here though, set the frequency early here but only enable the clock branch during the l2 startup sequence.